### PR TITLE
Deprecate Bio.Record.database_letters

### PR DESCRIPTION
**Pull request by souravsingh**
Fixes a part of #1000

### CODE REVIEW

File: Bio/Blast/Record.py [hunk 1]
```diff
@@ -54,7 +56,15 @@ def __init__(self):
...
-
+        
+        @property
+        def database_letters(self):
+            
+            warnings.warn("Accessing the .database_letters will "
+                          "be deprecated in the next " 
+                          "version of Biopython. ", BiopythonDeprecationWarning)
+            
+            return str(self)
```
Comment (peterjc) [inline Bio/Blast/Record.py]: That return value seems completely wrong - it should be an integer value. The tests ought to catch this...

The standard approach is to have the property proxy a private variable ``self._database_letters``, see e.g. https://github.com/biopython/biopython/blob/biopython-169/Bio/SeqFeature.py#L890

**Comment by peterjc** [comment]
I believe this approach will indirectly make the attribute read only, which might be OK but I am puzzled why this doesn't seem to break any of the unit tests.

Note there are some style issues - do you have the git hook installed to help catch this kind of thing earlier?

```
$ flake8 Bio/
Bio/Blast/Record.py:59:1: W293 blank line contains whitespace
Bio/Blast/Record.py:62:1: W293 blank line contains whitespace
Bio/Blast/Record.py:64:55: W291 trailing whitespace
Bio/Blast/Record.py:66:1: W293 blank line contains whitespace
Bio/Blast/Record.py:69:1: E302 expected 2 blank lines, found 1
```

**Comment by peterjc** [comment]
So should the class header have:

```python
self._database_letters = None
```

rather than:

```python
self.database_letters = None
```

**Comment by souravsingh** [comment]
I will make the changes. Do we have to write a docstring for the property?

**Comment by souravsingh** [comment]
@peterjc Is there any more updates needed, apart from the pep8 fixes?

**Comment by peterjc** [comment]
I would like to see a test added using the unittest frame work to confirm the parser can still set the values (this bit worries me) and that accessing the property raises the warning (for full test coverage of the new code).

**Comment by peterjc** [comment]
This example might be helpful for testing the warning is raised:
https://github.com/biopython/biopython/blob/biopython-169/Tests/test_GenBank_unittest.py#L57

**Comment by souravsingh** [comment]
We have to check for the values using the ``num_letters_in_database``, correct?

**Comment by souravsingh** [comment]
@peterjc Where do the test for deprecation need to be added?

**Comment by peterjc** [comment]
Probably ``test_NCBITextParser.py`` and ``test_NCBIXML.py`` as both uses the same objects.

**Comment by souravsingh** [comment]
@peterjc The test for Deprecation would be something like this, correct?
```python
import warnings`
from Bio import BiopythonDeprecationWarning
with warnings.catch_warnings():
    warnings.simplefilter("ignore", BiopythonDeprecationWarning)


```
`

**Comment by peterjc** [comment]
No, that would ignore the warning.  I want a test that the warning occurs. I was suggesting upgrading the warning to an error, then explicitly testing the warning/error occurs using either ``self.assertRaises`` or a try/except as in the example I linked to which checks the exact warning message: https://github.com/biopython/biopython/blob/biopython-169/Tests/test_GenBank_unittest.py#L57

File: Tests/test_NCBIXML.py [hunk 2]
```diff
@@ -1865,7 +1867,21 @@ def test_xml_2218L_rpsblast_001(self):
...
+            with open(os.path.join("Blast", filename)) as handle:
+                try:
+                    records = NCBIXML.parse(handle)
+                except BiopythonDeprecationWarning as de:
+                    self.assertEqual(str(e),
+                                     "Accessing the .database_letters will" 
+                                     "be deprecated in the next" 
+                                     "version of Biopython.")
+                else:
+                    elf.assertTrue(False, "Expected BiopythonDeprecationWarning here.")
```
Comment (peterjc) [inline Tests/test_NCBIXML.py]: Typo: ``self``

Do you have the git pre-commit hook installed? That should have spotted this...

File: Tests/test_NCBIXML.py [hunk 3]
```diff
@@ -1865,7 +1867,21 @@ def test_xml_2218L_rpsblast_001(self):
...
-
+    
+    def test_database_letters_deprecation(self):
+        with warnings.catch_warnings():
+            warnings.simplefilter("error", BiopythonDeprecationWarning)
+            filename = 'xml_2212L_blastp_001.xml'
+            with open(os.path.join("Blast", filename)) as handle:
+                try:
+                    records = NCBIXML.parse(handle)
+                except BiopythonDeprecationWarning as de:
```
Comment (peterjc) [inline Tests/test_NCBIXML.py]: Another typo ``de`` but in the next line you use ``e`` (short for error or exception).

PEP8 says using single letter variables is consider poor style, but ``e`` is commonly quite used in this situation.

File: Tests/test_NCBIXML.py (hunk 2 above)
Comment (souravsingh) [inline Tests/test_NCBIXML.py]: I don't. I will have to pull from the latest source.

File: Tests/test_NCBIXML.py [hunk 4]
```diff
@@ -1865,7 +1867,21 @@ def test_xml_2218L_rpsblast_001(self):
...
+                    self.assertEqual(str(e),
+                                     "Accessing the .database_letters will" 
+                                     "be deprecated in the next" 
+                                     "version of Biopython.")
```
Comment (peterjc) [inline Tests/test_NCBIXML.py]: Your string is missing spaces between "will" and "be", and "next" and "version". i.e. where you've split it up.

**Comment by peterjc** [comment]
It appears you have committed the latest changes and pushed them to github without running the tests locally. Running the test locally before pushing to github is a good habit to adopt.

**Comment by souravsingh** [comment]
Would it be ``True`` in this statement instead of ``False``-https://travis-ci.org/biopython/biopython/jobs/225743544#L633?

**Comment by peterjc** [comment]
The test seems to be correctly failing because the warning does not happen.

Have you actually tested this directly (by hand at the Python prompt)? i.e. Can you create a BLAST report object and try reading the ``.database_letters`` property? Do you see the new warning?

(I have not yet tried running your code myself yet)

**Comment by peterjc** [comment]
Also, separately from the coding and testing problem, again the TravisCI style check is failing:

```
$ flake8 Bio/
Bio/Blast/Record.py:59:1: W293 blank line contains whitespace
Bio/Blast/Record.py:63:1: W293 blank line contains whitespace
Bio/Blast/Record.py:65:63: W291 trailing whitespace
Bio/Blast/Record.py:66:1: W293 blank line contains whitespace
Bio/Blast/Record.py:70:1: W293 blank line contains whitespace
Bio/Blast/Record.py:72:55: W291 trailing whitespace
Bio/Blast/Record.py:74:1: W293 blank line contains whitespace
```

You can install and run ``flake8`` on your machine (``pip install flake8``), but as mentioned before I would urge you to try the git pre-commit hook (see the ``CONTRIBUTING.rst`` file which points at https://github.com/biopython/biopython/issues/493 for details).

**Comment by souravsingh** [comment]
@peterjc ``flake8```doesn't detect any of the above problems when running on the machine.

**Comment by peterjc** [comment]
Perhaps your flake8 is out of date? I just cloned your repository, checkout out this branch and get:

```
$ flake8 --version
3.3.0 (mccabe: 0.6.1, pycodestyle: 2.3.1, pyflakes: 1.5.0) CPython 3.6.0 on Darwin
$ flake8 Bio/
Bio/Blast/Record.py:59:1: W293 blank line contains whitespace
Bio/Blast/Record.py:63:1: W293 blank line contains whitespace
Bio/Blast/Record.py:65:63: W291 trailing whitespace
Bio/Blast/Record.py:66:1: W293 blank line contains whitespace
Bio/Blast/Record.py:70:1: W293 blank line contains whitespace
Bio/Blast/Record.py:72:55: W291 trailing whitespace
Bio/Blast/Record.py:74:1: W293 blank line contains whitespace
```

Try ``pip install --upgrade flake8`` if your flake is older than v3.3.0

**Comment by peterjc** [comment]
Good progress - it looks like you must have got ``flake8`` working locally, and got ``Bio/Blast/Record.py`` fixed up 👍 

You'll also need to do ``Tests/test_NCBIXML.py`` to make our TravisCI style check pass.

```
$ flake8 Tests/
Tests/test_NCBIXML.py:1870:1: W293 blank line contains whitespace
Tests/test_NCBIXML.py:1880:77: W291 trailing whitespace
Tests/test_NCBIXML.py:1881:66: W291 trailing whitespace
Tests/test_NCBIXML.py:1886:1: E305 expected 2 blank lines after class or function definition, found 1
```

**Comment by souravsingh** [comment]
I didn't realise I was using an outdated version of ``flake8``, due to which I wasn't getting the required errors. Updating ``flake8`` now worked well for me.

**Comment by souravsingh** [comment]
@peterjc The style checks now pass completely. Are there any other fixes that need to be done?

**Comment by peterjc** [comment]
Good work, yes, the TravisCI style check passes now. However, the actual tests are all failing, e.g.

```
======================================================================
FAIL: test_database_letters_deprecation (test_NCBIXML.TestNCBIXML)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/biopython/biopython/Tests/test_NCBIXML.py", line 2783, in test_database_letters_deprecation
    False, "Expected BiopythonDeprecationWarning here.")
AssertionError: Expected BiopythonDeprecationWarning here.
----------------------------------------------------------------------
```

Once again, have you actually tested the new warning directly (by hand at the Python prompt)? i.e. Can you create a BLAST report object and try reading the ``.database_letters`` property? Do you see the new warning?

**Comment by souravsingh** [comment]
I am getting an AttributeError

```python
>>> for record in NCBIXML.parse(open("xml_blastp.xml")):
...     print record.query_id
...     print record.database_length
...     print record.database_letters
... 
gi|49176427|ref|NP_418280.3|
1011751523
Traceback (most recent call last):
  File "<stdin>", line 4, in <module>
AttributeError: 'Blast' object has no attribute 'database_letters'
```

**Comment by peterjc** [comment]
That seems odd. Testing on the current Biopython master branch,

```
$ cd biopython/Tests/Blast/
$ python3
Python 3.6.0 (v3.6.0:41df79263a11, Dec 22 2016, 17:23:13) 
[GCC 4.2.1 (Apple Inc. build 5666) (dot 3)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from Bio.Blast import NCBIXML 
>>> for record in NCBIXML.parse(open("xml_2226_tblastx_001.xml")):
...     print(record.query_id)
...     print(record.database_length)
...     print(record.database_letters)
... 
Query_1
67750
None
Query_2
67750
None
```

Testing the objects directly.

```
$ python3
fPython 3.6.0 (v3.6.0:41df79263a11, Dec 22 2016, 17:23:13) 
[GCC 4.2.1 (Apple Inc. build 5666) (dot 3)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from Bio.Blast.Record import Header
>>> h = Header()
>>> dir(h)
['__class__', '__delattr__', '__dict__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__le__', '__lt__', '__module__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', '__weakref__', 'application', 'database', 'database_letters', 'database_sequences', 'date', 'query', 'query_letters', 'reference', 'version']
>>> h.database_letters
>>> repr(h.database_letters)
'None'
```

Also the ``Blast`` object subclasses ``Header``, so it has it too:

```
>>> from Bio.Blast.Record import Blast
>>> b = Blast()
>>> dir(b)
['__class__', '__delattr__', '__dict__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__le__', '__lt__', '__module__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', '__weakref__', 'alignments', 'application', 'blast_cutoff', 'database', 'database_length', 'database_letters', 'database_name', 'database_sequences', 'date', 'descriptions', 'dropoff_1st_pass', 'effective_database_length', 'effective_hsp_length', 'effective_query_length', 'effective_search_space', 'effective_search_space_used', 'frameshift', 'gap_penalties', 'gap_trigger', 'gap_x_dropoff', 'gap_x_dropoff_final', 'gapped', 'hsps_gapped', 'hsps_no_gap', 'hsps_prelim_gapped', 'hsps_prelim_gapped_attemped', 'ka_params', 'ka_params_gap', 'matrix', 'multiple_alignment', 'num_good_extends', 'num_hits', 'num_letters_in_database', 'num_seqs_better_e', 'num_sequences', 'num_sequences_in_database', 'posted_date', 'query', 'query_id', 'query_length', 'query_letters', 'reference', 'sc_match', 'sc_mismatch', 'threshold', 'version', 'window_size']
>>> b.database_letters
>>> repr(b.database_letters)
'None'
```

**Comment by souravsingh** [comment]
I had built and installed Biopython from the feature branch in a virtual environment before proceeding to test the property.

**Comment by peterjc** [comment]
It look me a while to understand why your code change was doing this. It was a learning experience.

The ``AttributeError`` is from Python telling you that the class does not have a ``database_letters`` attribute. This means your property definition did not work... but why?

Hint: Check your indentation.

**Comment by souravsingh** [comment]
@peterjc I have made the fix and now I get the required Deprecation Warning-

```python
Python 2.7.12 (default, Nov 19 2016, 06:48:10) 
[GCC 5.4.0 20160609] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> from Bio.Blast import NCBIXML
>>> for record in NCBIXML.parse(open("xml_blastp.xml")):
...     print record.query_id
...     print record.database_length
...     print record.database_letters
... 
gi|49176427|ref|NP_418280.3|
1011751523
/home/sourav/myenv-bio/lib/python2.7/site-packages/Bio/Blast/Record.py:75: BiopythonDeprecationWarning: Accessing the .database_letters will be deprecated in the next version of Biopython .
  BiopythonDeprecationWarning)
None
>>> 
```

**Comment by peterjc** [comment]
Good, yes - please fix the indentation on this branch. Have you run the test suite on your machine yet?

Can you guess what I'm going to ask you to check/fix next?

**Comment by souravsingh** [comment]
I have run the test suite on the machine. One failure-
```shell
======================================================================
FAIL: test_database_letters_deprecation (__main__.TestNCBIXML)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_NCBIXML.py", line 2783, in test_database_letters_deprecation
    False, "Expected BiopythonDeprecationWarning here.")
AssertionError: Expected BiopythonDeprecationWarning here.

```

As for further fixes, I am thinking that another unit-test needs to be added to check the values of ``database_letters`` and a note in ``DEPRECATED``?

**Comment by peterjc** [comment]
First, an easy point - the warning message wording: If we wanted this to be just a warning of a future (pending) deprecation, Python has a ``PendingDeprecationWarning``. However, the point of this change is to actually deprecate the property (using ``BiopythonDeprecationWarning``), and the message should tell the user what to do instead. Please refer to http://biopython.org/wiki/Deprecation_policy and the (long) example warning there for help with wording this.

Second: What happens when code tries to set this value? e.g. The BLAST parsers? Does it work? If not, what needs changing?

**Comment by peterjc** [comment]
In answer to your questions, there ought to be unit tests already checking the value of ``.database_letters``, but please do confirm than and if not, add this.

And yes, this would deserve a mention in the `DEPRECATED file.

**Comment by souravsingh** [comment]
So what do we do about the test failure, Do we add a filter to ignore the error?

**Comment by peterjc** [comment]
OK, as I expected, we have a new set of test failures like this:

```
======================================================================
ERROR: test_conversion (test_NCBITextParser.TestBlastRecord)
Test converting a Blast record multiple alignment
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/biopython/biopython/Tests/test_NCBITextParser.py", line 30, in test_conversion
    record = self.parser.parse(handle)
  File "/home/travis/build/biopython/biopython/build/lib.linux-x86_64-2.7/Bio/Blast/NCBIStandalone.py", line 811, in parse
    self._scanner.feed(handle, self._consumer)
  File "/home/travis/build/biopython/biopython/build/lib.linux-x86_64-2.7/Bio/Blast/NCBIStandalone.py", line 98, in feed
    self._scan_header(uhandle, consumer)
  File "/home/travis/build/biopython/biopython/build/lib.linux-x86_64-2.7/Bio/Blast/NCBIStandalone.py", line 202, in _scan_header
    read_and_call(uhandle, consumer.database_info, contains='sequences')
  File "/home/travis/build/biopython/biopython/build/lib.linux-x86_64-2.7/Bio/ParserSupport.py", line 131, in read_and_call
    method(line)
  File "/home/travis/build/biopython/biopython/build/lib.linux-x86_64-2.7/Bio/Blast/NCBIStandalone.py", line 877, in database_info
    self._header.database_letters = _safe_int(letters)
AttributeError: can't set attribute
======================================================================
```

This is why I asked yesterday:

>  What happens when code tries to set this value? e.g. The BLAST parsers? Does it work? If not, what needs changing?

Hint: Before ``.database_letters`` was a normal attribute which could be read and written to. You've replaced it with a read only property.

File: Bio/Blast/Record.py [hunk 5]
```diff
@@ -74,7 +74,7 @@ def database_letters(self):
             "version of Biopython .",
             BiopythonDeprecationWarning)
 
-        return self._database_letters
+        return self.database_letters
```
Comment (peterjc) [inline Bio/Blast/Record.py]: This change defines the property ``.database_letters`` in terms of itself, which is likely to cause a recursion error (or some other error - I'm curious to see exactly what will happen).

Comment (souravsingh) [inline Bio/Blast/Record.py]: It showing the same AttributeError

Comment (peterjc) [inline Bio/Blast/Record.py]: OK, not the exact error I expected - but you should still revert this change.

**Comment by peterjc** [comment]
The rigorous way to fix this is to define a read/write property.

However, as it seems unlikely any other code would need to set ``.database_letters``, I have a potential one-line fix in mind...

Have you checked to see where ``.database_letters`` is set in the code? i.e. Where do we attempt to set this attribute/property's value? Hint: The BLAST parsing code. Try using ``grep`` (or the GitHub search) to look at all the Python code.

**Comment by souravsingh** [comment]
Doing a ``git grep`` for all instances of ``database_letters`` shows this output-

```shell
Bio/Blast/NCBIStandalone.py:            self._header.database_letters = _safe_int(letters)
Bio/Blast/NCBIXML.py:        # TODO? Deprecate database_letters next?
Bio/Blast/Record.py:    database_letters    Number of letters in the database.  (int)
Bio/Blast/Record.py:        self.database_letters = None
Tests/test_NCBITextParser.py:        self.assertEqual(record.database_letters, 29652561)
Tests/test_NCBITextParser.py:        self.assertEqual(record.database_letters, 1461753)
Tests/test_NCBITextParser.py:        self.assertEqual(record.database_letters, 29652561)
Tests/test_NCBITextParser.py:        self.assertEqual(record.database_letters, 29652561)
Tests/test_NCBITextParser.py:        self.assertEqual(record.database_letters, 29652561)
Tests/test_NCBITextParser.py:        self.assertEqual(record.database_letters, 29652561)
Tests/test_NCBITextParser.py:        self.assertEqual(record.database_letters, 29652561)
Tests/test_NCBITextParser.py:        self.assertEqual(record.database_letters, 29652561)
Tests/test_NCBITextParser.py:        self.assertEqual(record.database_letters, 31998854)
Tests/test_NCBITextParser.py:        self.assertEqual(record.database_letters, 31998854)
Tests/test_NCBITextParser.py:        self.assertEqual(record.database_letters, 31998854)
Tests/test_NCBITextParser.py:        self.assertEqual(record.database_letters, 31998854)
Tests/test_NCBITextParser.py:        self.assertEqual(record.database_letters, 29652561)
Tests/test_NCBITextParser.py:        self.assertEqual(record.database_letters, 29652561)
Tests/test_NCBITextParser.py:        self.assertEqual(record.database_letters, 31998854)
Tests/test_NCBITextParser.py:        self.assertEqual(record.database_letters, 31998854)
Tests/test_NCBITextParser.py:        self.assertEqual(record.database_letters, 31998854)
Tests/test_NCBITextParser.py:        self.assertEqual(record.database_letters, 29652561)
Tests/test_NCBITextParser.py:        self.assertEqual(record.database_letters, 29652561)
Tests/test_NCBITextParser.py:        self.assertEqual(record.database_letters, 29652561)
Tests/test_NCBITextParser.py:        self.assertEqual(record.database_letters, 1461753)
Tests/test_NCBITextParser.py:        self.assertEqual(record.database_letters, 29652561)
Tests/test_NCBITextParser.py:        self.assertEqual(record.database_letters, 29652561)
Tests/test_NCBITextParser.py:        self.assertEqual(record.database_letters, 29652561)
Tests/test_NCBITextParser.py:        self.assertEqual(record.database_letters, 29652561)
Tests/test_NCBITextParser.py:        self.assertEqual(record.database_letters, 29652561)
Tests/test_NCBITextParser.py:        self.assertEqual(record.database_letters, 31998854)
Tests/test_NCBITextParser.py:        self.assertEqual(record.database_letters, 31998854)
Tests/test_NCBITextParser.py:        self.assertEqual(record.database_letters, 31998854)
Tests/test_NCBITextParser.py:        self.assertEqual(record.database_letters, 31998854)
Tests/test_NCBITextParser.py:        self.assertEqual(record.database_letters, 29652561)
Tests/test_NCBITextParser.py:        self.assertEqual(record.database_letters, 29652561)
Tests/test_NCBITextParser.py:        self.assertEqual(record.database_letters, 31998854)
Tests/test_NCBITextParser.py:        self.assertEqual(record.database_letters, 31998854)
Tests/test_NCBITextParser.py:        self.assertEqual(record.database_letters, 31998854)
Tests/test_NCBITextParser.py:        self.assertEqual(record.database_letters, 29652561)
Tests/test_NCBITextParser.py:        self.assertEqual(record.database_letters, 29652561)
Tests/test_NCBITextParser.py:        self.assertEqual(record.database_letters, 29652561)
Tests/test_NCBITextParser.py:        self.assertEqual(record.database_letters, 31957340)
Tests/test_NCBITextParser.py:        self.assertEqual(record.database_letters, 31998854)
Tests/test_NCBITextParser.py:        self.assertEqual(record.database_letters, 31753758)
Tests/test_NCBITextParser.py:        self.assertEqual(record.database_letters, 7312)
Tests/test_NCBITextParser.py:        self.assertEqual(record.database_letters, 4662239)
Tests/test_NCBITextParser.py:        self.assertEqual(record.database_letters, 1221421)
Tests/test_NCBITextParser.py:        self.assertEqual(record.database_letters, 1221421)
Tests/test_NCBITextParser.py:        self.assertEqual(record.database_letters, 4214630)
Tests/test_NCBITextParser.py:        self.assertEqual(record.database_letters, 4214630)
Tests/test_NCBITextParser.py:        self.assertEqual(record.database_letters, 1350842)
Tests/test_NCBITextParser.py:        self.assertEqual(record.database_letters, 1573298872)
Tests/test_NCBITextParser.py:        self.assertEqual(record.database_letters, 3078807967)
Tests/test_NCBITextParser.py:        self.assertEqual(record.database_letters, 3078807967)
Tests/test_NCBITextParser.py:        self.assertEqual(record.database_letters, 3078807967)
Tests/test_NCBITextParser.py:        self.assertEqual(record.database_letters, 3078807967)
Tests/test_NCBITextParser.py:        self.assertEqual(record.database_letters, 3078807967)
Tests/test_NCBITextParser.py:        self.assertEqual(record.database_letters, 3078807967)
Tests/test_NCBITextParser.py:        self.assertEqual(record.database_letters, 3078807967)
Tests/test_NCBITextParser.py:        self.assertEqual(record.database_letters, 3078807967)
Tests/test_NCBITextParser.py:        self.assertEqual(record.database_letters, 3078807967)
Tests/test_NCBITextParser.py:        self.assertEqual(record.database_letters, 3078807967)
Tests/test_NCBITextParser.py:        self.assertEqual(record.database_letters, 3078807967)
Tests/test_NCBITextParser.py:        self.assertEqual(record.database_letters, 3078807967)
Tests/test_NCBITextParser.py:        self.assertEqual(record.database_letters, 3078807967)
Tests/test_NCBITextParser.py:        self.assertEqual(record.database_letters, 3078807967)
Tests/test_NCBITextParser.py:        self.assertEqual(record.database_letters, 3078807967)
Tests/test_NCBITextParser.py:        self.assertEqual(record.database_letters, 3078807967)
Tests/test_NCBITextParser.py:        self.assertEqual(record.database_letters, 3078807967)
Tests/test_NCBITextParser.py:        self.assertEqual(record.database_letters, 3078807967)
Tests/test_NCBITextParser.py:        self.assertEqual(record.database_letters, 3078807967)
Tests/test_NCBITextParser.py:        self.assertEqual(record.database_letters, 3078807967)
Tests/test_NCBITextParser.py:        self.assertEqual(record.database_letters, 3078807967)
Tests/test_NCBITextParser.py:        self.assertEqual(record.database_letters, 3078807967)
Tests/test_NCBITextParser.py:        self.assertEqual(record.database_letters, 3078807967)
Tests/test_NCBITextParser.py:        self.assertEqual(record.database_letters, 3078807967)
Tests/test_NCBITextParser.py:        self.assertEqual(record.database_letters, 3078807967)
Tests/test_NCBITextParser.py:        self.assertEqual(record.database_letters, 4626651242)
Tests/test_NCBITextParser.py:        self.assertEqual(record.database_letters, 4626651242)
Tests/test_NCBITextParser.py:        self.assertEqual(record.database_letters, 4626651242)
Tests/test_NCBITextParser.py:        self.assertEqual(record.database_letters, 4140237112)
Tests/test_NCBITextParser.py:        self.assertEqual(record.database_letters, 4140237112)
Tests/test_NCBITextParser.py:        self.assertEqual(record.database_letters, 4140237112)
Tests/test_NCBITextParser.py:        self.assertEqual(record.database_letters, 4140237112)
Tests/test_NCBITextParser.py:        self.assertEqual(record.database_letters, 4140237112)
Tests/test_NCBITextParser.py:        self.assertEqual(record.database_letters, 4140237112)
Tests/test_NCBITextParser.py:        self.assertEqual(record.database_letters, 4626651242)
Tests/test_NCBITextParser.py:        self.assertEqual(record.database_letters, 4626651242)
Tests/test_NCBITextParser.py:        self.assertEqual(record.database_letters, 4626651242)
Tests/test_NCBITextParser.py:        self.assertEqual(record.database_letters, 4626651242)
Tests/test_NCBITextParser.py:        self.assertEqual(record.database_letters, 4626651242)
Tests/test_NCBITextParser.py:        self.assertEqual(record.database_letters, 4626651242)
```

**Comment by peterjc** [comment]
Yes, most are tests comparing the value. But where do we try to set the value? Where were we getting the exception ``AttributeError: can't set attribute`` before?

**Comment by souravsingh** [comment]
All the AttributeErrors are arising from ``test_NCBITextParser.py``

**Comment by peterjc** [comment]
No. We do not need to change the existing tests - they are working perfectly. We need to make sure this pull request does not *break* the existing tests.

I've been trying not to simply tell you what the problem is and how to fix it, but guide you to solving it yourself. I may have misjudged your knowledge of Python, and I'm sorry if thus has been frustrating.

Before ``.database_letters`` was a normal attribute which could be read and written to. Our BLAST parsers write to it to record the number of letters in the database. Scripts using the parser and our tests read the value from this attribute.

In order to add a warning, we have to convert this into a property where a bit of cade gets called when the property is read from or set. Right now you've replaced it with a read only property, which means when our parser tries to record the number of letters in the database it can't:

```
Bio/Blast/NCBIStandalone.py", line 877, in database_info
    self._header.database_letters = _safe_int(letters)
AttributeError: can't set attribute
```

One way to fix this is to define two methods, on for getting the value (done), and one for setting the value (missing). Both could give warnings. You can find examples of doing this the old-fashioned way elsewhere in Biopython using ``property(...)`` but this is now possible using decorators instead.

The simpler fix I had in mind was having our parser directly set the value of the new private variable 
``._database_letters`` instead, i.e. ``Bio/Blast/NCBIStandalone.py`` line 877 would becomes:

```python
    self._header.database_letters = _safe_int(letters)
```

**Comment by souravsingh** [comment]
I thought the Deprecation would be an easier process. Damn!

**Comment by souravsingh** [comment]
The code is already there in ``NCBIStandalone.py``- https://github.com/biopython/biopython/blob/master/Bio/Blast/NCBIStandalone.py#L877

**Comment by souravsingh** [comment]
Are there any other nitfixes required?

**Comment by peterjc** [comment]
Can you revert 28780c999342aeccafccbd903c40b339339555f2 which seems to have massively changed the test file - probably most of it is harmless white space but it makes it very hard to review the actual changes on your pull request.

**Comment by souravsingh** [comment]
Revert is done

File: Bio/Blast/Record.py [hunk 6]
```diff
@@ -53,19 +56,43 @@ def __init__(self):
 
         self.database = ''
         self.database_sequences = None
-        self.database_letters = None
+        self._database_letters = None
+
+    @property
+    def database_letters(self):
+        """Database length info of the plain text BLAST format(DEPRECATED).
```
Comment (peterjc) [inline Bio/Blast/Record.py]: Needs a space before ``(DEPRECATED).``

File: Bio/Blast/Record.py [hunk 7]
```diff
@@ -53,19 +56,43 @@ def __init__(self):
...
+
+        This is a read only property provided by Biopython to access the
```
Comment (peterjc) [inline Bio/Blast/Record.py]: You've removed the read-only limitation now.

File: Bio/Blast/Record.py [hunk 8]
```diff
@@ -53,19 +56,43 @@ def __init__(self):
...
+        actual database size from a plain text BLAST file.
+
+        It is preferred that users use num_letters_in_database from
+        Bio.Record.DatabaseReport instead of database_letters.
+        """
+        warnings.warn(
```
Comment (peterjc) [inline Bio/Blast/Record.py]: Please update the warning tense (your change is deprecating it) as previously noted. You will also need to update the test to match.

**Comment by peterjc** [comment]
You have a new test failure,

```
======================================================================
FAIL: test_database_letters_deprecation (test_NCBIXML.TestNCBIXML)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/biopython/biopython/Tests/test_NCBIXML.py", line 1884, in test_database_letters_deprecation
    self.assertTrue(False, "Expected BiopythonDeprecationWarning here.")
AssertionError: Expected BiopythonDeprecationWarning here.
----------------------------------------------------------------------
```

**Comment by souravsingh** [comment]
The test failure is coming from the test which is written for Deprecation.

**Comment by souravsingh** [comment]
I have added the entry to ``DEPRECATED``

**Comment by souravsingh** [comment]
@peterjc Are there any more fixes needed for the PR...?

**Comment by peterjc** [comment]
Given current work on the master branch to tighten our style checks, it is understandable if that stops passing.

However, what is a real problem is the code tests are failing, eg:

```
======================================================================
FAIL: test_database_letters_deprecation (test_NCBIXML.TestNCBIXML)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/biopython/biopython/Tests/test_NCBIXML.py", line 1883, in test_database_letters_deprecation
    self.assertTrue(False, "Expected BiopythonDeprecationWarning here.")
AssertionError: Expected BiopythonDeprecationWarning here.
----------------------------------------------------------------------
```

**Comment by souravsingh** [comment]
@peterjc I believe you mentioned regarding the test failure for deprecation-

> The test seems to be correctly failing because the warning does not happen.

**Comment by peterjc** [comment]
That comment of mine was in the context of your code not actually giving the warning.

We want the warning to happen, and the test should confirm the warning is triggered.

**Comment by souravsingh** [comment]
what would be the fix for this?

**Comment by souravsingh** [comment]
Ping @peterjc

**Comment by peterjc** [comment]
Just from looking at the code on GitHub, I'm not sure. I would have to try it on my own machine. I would start with the kind of checks I did back in my earlier comment https://github.com/biopython/biopython/pull/1181#issuecomment-297688898 and also attempt to set the value as well.

**Comment by souravsingh** [comment]
I have tested the getter object and it shows the Deprecation Warning. I will take a look at the setter object.

**Comment by souravsingh** [comment]
@peterjc Would it be possible to have the patch merged before 1.70?

**Comment by peterjc** [comment]
If you can get the tests to pass, sure.

**Comment by souravsingh** [comment]
Closing the PR since there is no solution to resolve the issue at hand.
